### ISSUE TITLE
Replacing a Genome

### ISSUE THREAD

**Issue by francicco** [ComparativeGenomicsToolkit/cactus#1033]
Hi @glennhickey,

I have a WGA of 63 lepidoptera species. I would like to update the alignment instead of starting from scratch. The easiest thing would be to replace a genome. I'm trying to look at the tutorial but I found it confusing. What I did was:

- Removing the genome that I want to replace with `halRemoveGenome` (easy)
- I create a config file with ALL the existing genome + the path for the new one, but then I'm confused. Do I have to use `halReplaceGenome` or re run `cacus`

Can you help me please?
F

P.S. at some point I'd want to add many more species exactly where the species I want to replace.

**Comment by glennhickey** [comment]
Have you looked at [these instructions?](https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/cactus-update-prepare.md#replacing-a-genome)

**Comment by francicco** [comment]
Also, for some reason `cactus-update-prepare` get stuck at `halValidate`
F

**Comment by francicco** [comment]
Hi Glenn,

Thanks a lot for these tutorials, very helpful!

I'm trying to combine two hal like this:

`halAppendSubtree 63Nymphalidae.Cactus-v1.5.hal Danaiinae.hal Danaiinae Danaiinae --merge --hdf5InMemory --noMarkAncestors`

But I'm getting the following error:

```
[halAppendSubtree] Copying DanAnc01
HDF5-DIAG: Error detected in HDF5 (1.10.9) thread 0:
  #000: H5Dio.c line 186 in H5Dread(): can't read data
    major: Dataset
    minor: Read failed
  #001: H5Dio.c line 583 in H5D__read(): can't read data
    major: Dataset
    minor: Read failed
  #002: H5Dchunk.c line 2562 in H5D__chunk_read(): unable to read raw data chunk
    major: Low-level I/O
    minor: Read failed
  #003: H5Dchunk.c line 3930 in H5D__chunk_lock(): data pipeline read failed
    major: Dataset
    minor: Filter operation failed
  #004: H5Z.c line 1390 in H5Z_pipeline(): filter returned failure during read
    major: Data filters
    minor: Read failed
  #005: H5Zdeflate.c line 118 in H5Z__filter_deflate(): inflate() failed
    major: Data filters
    minor: Unable to initialize object
terminate called after throwing an instance of 'H5::DataSetIException'
Aborted
```

Any help?
Cheers
F

**Comment by francicco** [comment]
Hi @glennhickey 

I fix it, the recipient hal got corrupted. I repeated it and it went smoothly.
I have another question. 

Attaching different hal files works really well and it's a clever way to add new genomes.  I noticed that the final phylogenetic tree within the hal suffers a bit because the branches across the tree is obviously non-homogeneous. How this is affecting the alignment itself, and how to fix it, if there's a way.

<img width="780" alt="Screenshot 2023-06-03 at 21 07 51" src="https://github.com/ComparativeGenomicsToolkit/cactus/assets/9006870/f1999001-5fea-43c3-8ea0-3b8d0d2b1326">

Thanks a lot and sorry for bothering you so many times!
Best
F

**Comment by glennhickey** [comment]
There's a bit of explanation of how branch lengths are used during alignment [in the faq](https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/progressive.md#frequently-asked-questions). 

Given that your alignment is already done, there's not much you can do except for re-align using the corrected branch lengths. But as explained above, I don't think having too-long branch lengths will have too much of an effect on the alignment itself.

**Comment by francicco** [comment]
Hi Glenn,

I'm having a hard time trying to update some genomes in my alignment. Sometimes it works sometimes it really doesn't and I don't understand why. I don't know what I'm doing wrong.
What I want to do is replace a subtree. This is what I'm doing:
- I'm removing the subtree with `halRemoveSubtree`; eg: (A,B)C;
- I extract the root of the subtree that I want to plug in "C";
- I execute `cactus-update-prepare replace MAINHAL ReplaceC.txt --genome C --outDir ./C.steps --jobStore ./C.jobstore`
- This generates a set of commands that I execute. They all run without errors, with the exaption of `halReplaceGenome --bottomAlignmentFile ./C.steps/NewMRCA.hal --topAlignmentFile MAINHAL MAINHAL NewMRCA --hdf5InMemory`
`terminate called after throwing an instance of 'hal_exception'
  what():  Trying to set top segment coordinate out of range`

Also at this point MAINHAL is corrupted, I guess.

I used the procedure in a previous hal update and it worked perfectly.
Can you help me understand?

Thanks a lot
F

**Comment by glennhickey** [comment]
Hi, I wish I could say from here what's going wrong but I can't .  I think there have been other similar issues.  If you are somehow able to share the input data, I will be able to debug here.  The tools to update alignments can be very frustrating, but I haven't yet run into any situations like this yet when using them myself.

**Comment by francicco** [comment]
I can upload it on google drive, quite big files but I should be able. What files do you need?
I'd also need an email I guess
F

**Comment by glennhickey** [comment]
I guess we can start with the input to `halReplaceGenome`.  If the alignment you're replacing in is very large, maybe you can use `halExtractRoot` to prune off the subtree you're trying to replace into?  My github name at gmail will work.

**Comment by francicco** [comment]
Hi @glennhickey 

Did you see my email?
Cheers
F

**Comment by glennhickey** [comment]
I did not.  But I just found it in my Spam folder.  Will take a look.

**Comment by glennhickey** [comment]
Just to confirm, can you please send the exact command (on the files you shared) that your were running?

**Comment by francicco** [comment]
```
## Alignment update
halReplaceGenome --bottomAlignmentFile ./Argynnini.steps/Heliconiinae.hal --topAlignmentFile 75way-Nymphalidae.Cactus-v1.0.hal 75way-Nymphalidae.Cactus-v1.0.hal Heliconiinae --hdf5InMemory 
```

Thanks a lot for looking into this!
F

**Comment by glennhickey** [comment]
OK, thanks.  So I have these files:
```
75way-Nymphalidae.Cactus-v1.0.hal  md5sum: 7221a91d89e53f9a6cf59d7f84a2b1ec
Heliconiinae.hal                   md5sum: c37b221ab5e403450ed4823b6308f0e0
```
Then I run your command
```
halReplaceGenome --bottomAlignmentFile Heliconiinae.hal --topAlignmentFile 75way-Nymphalidae.Cactus-v1.0.hal 75way-Nymphalidae.Cactus-v1.0.hal Heliconiinae --hdf5InMemory
```
And I get an error too (but somehow more informative?)
```
terminate called after throwing an instance of 'hal_exception'
  what():  Cannot update sequence ArgynninirefChr0 because it is not present in  genome Argynnini
Aborted (core dumped)
```
Given that's the only genome that's different, I imagine that's the one we're updating. 

In the [documentation](https://github.com/ComparativeGenomicsToolkit/cactus/blob/master/doc/updating-alignments.md#replacing-a-genome) it says

> A genome can be replaced (for example to update an assembly version) by removing it and then following the "add-to-node" procedure to add the new version back as a child of its parent.

So let's try removing it first then repeating the `halReplaceGenome` command
```
cp bak/75way-Nymphalidae.Cactus-v1.0.hal 75way-Nymphalidae.Cactus-v1.0.hal 
halRemoveGenome 75way-Nymphalidae.Cactus-v1.0.hal Argynnini
halReplaceGenome --bottomAlignmentFile Heliconiinae.hal --topAlignmentFile 75way-Nymphalidae.Cactus-v1.0.hal 75way-Nymphalidae.Cactus-v1.0.hal Heliconiinae --hdf5InMemory
```
But that crashes too
```
Segmentation fault (core dumped)
```

This crash seems to trace back to a [relatively recent change I made to support bigger files](https://github.com/ComparativeGenomicsToolkit/hal/commit/c3f4a6c626e75be5ec25daa1800d78d2ba3e3666), which I will get fixed asap.  In the meantime, you should be able to work around it by using an older `halReplaceGenome`, like that from [v2.2.3](https://github.com/ComparativeGenomicsToolkit/cactus/releases/tag/v2.2.3).

So in summary, there are two bugs
* `cactus-update-prepare replace` omits the `halRemoveGenome` command
* `halReplaceGenome` itself is broken in the latest version of hal

But you can work around them now by adding the `halRemoveGenome` command yourself and running `halReplaceGenome` from an older release. 

thanks for your patience, and for sharing the data to get to the bottom of this.

**Comment by francicco** [comment]
Hi @glennhickey 

Thanks a lot for helping me with that! It worked! Now that I substitute the branch need to append its subtree, and it worked too!!!!
Thanks again to help me in my journey to extend my alignment! The next steps to learn would be to add branches at the root and in-between nodes! :D

Cheers
F

**Comment by John-Neddermeyer** [comment]
Hi @glennhickey 

I'm running into a similar issue where I get a segmentation fault error when running the command `singularity run cactus.v2.8.2.sif halReplaceGenome --bottomAlignmentFile steps_dir/Anc28.hal --topAlignmentFile ./cactus_multi_species_alignment.hal ./cactus_multi_species_alignment.hal Anc28 --hdf5InMemory`

I used `cactus-update-prepare replace` to determine the steps to updating an existing cactus alignment with a new genome version for a species. All other steps complete successfully `cactus-preprocess`, `cactus-blast`, `cactus-align`, and `halRemoveGenome`. I've used cactus versions v2.8.2, v2.6.8 and v2.2.3 with memory as high as 150GB. The replace genome step runs for at most 1 minute before returning the segmentation fault and `halReplaceGenome` is modifying the hal file before the error is thrown.

Thank you for your time.
John

**Comment by marcopessoa** [comment]
Has this been fixed? Had the same issue with 2.9.0

**Comment by John-Neddermeyer** [comment]
Not to my knowledge. I'm in the process of generating the alignment again from scratch. I'm not sure if the issue is related to updating alignments with highly similar genomes, but the step keeps failing when updating species with really short internal branches.
![Screenshot 2024-08-15 at 11 09 56 AM](https://github.com/user-attachments/assets/05189b28-0ae7-4620-ad6b-fa161d637eb5)

**Comment by glennhickey** [comment]
If anyone can share some input data that's crashing, I'll take a look.

**Comment by Xacfran** [comment]
I ran into the same issue when attempting to run `halReplaceGenome`. Just like other users, I have followed the steps given by `cactus-update-prepare` by following the adding-to-node approach, and have also tried different to assign 4 Gb - 9 Gb of memory but I keep getting segmentation fault and the job is killed.
Since the HAL file I am working with is ~250 Gb and has 113 terminals, I wanted to try using `halExtractRoot` to email you the subtree I am attempting to work with. However, the function is not available in the docker container of the latest version. Was it renamed? or is there a better way to send you a portion of the file for you to troubleshoot? 

I was also wondering if trying the approach defined [here](https://cgl.gi.ucsc.edu/data/cactus/241-mammalian-2020v2.1.README.md) to update and add canine genomes, could also be a way to go about it? I was thinking about removing the subtree with `halRemoveSubtree` and then append the new subtree alignment, but I am concerned that I run into the same error. Thanks a lot!
